### DISCUSSION TITLE
Preact support?

### DISCUSSION THREAD

**Discussion by vio** [ariakit/ariakit#2690] [Questions] [answered]
Hi! 

Curious to know if you consider supporting Preact with `@ariakit/react`?

## Context

I am maintaining https://github.com/relative-ci/bundle-stats, and the [standalone report](https://relative-ci.com/tools/webpack-bundle-stats/demo-multiple-jobs.html) is a Preact(React & preact/compat) SPA. The application has been using `reakit@1x` and `ariakit@2.0.0-next` for some of the components.  It has been working great, though atm we are stuck at `ariakit@2.0.0-next.44` and `preact@10.10.x`

- on `preact@10.11.x`, the tooltip/hovercard position is not set
- `@ariakit/react` throws `Uncaught Error: Cannot call an event handler while rendering` - https://stackblitz.com/edit/gkmbyl-fb7xyb?file=vite.config.ts

---

@diegohaz Thank you for the fantastic work you put into reakit/ariakit 🙇

**Comment by diegohaz** [comment, marked as answer]
I don't have plans to provide specific support for Preact. To do so, we'd need to have automated tests for Preact, which we've never had. If it's worked in the past, it's due to Preact's compatibility with React. So, if it's not working now, it's also due to Preact's compatibility with React.

Ariakit uses React's `useIntersectionEffect` hook when it's available: https://github.com/ariakit/ariakit/blob/f41c0f9134a4714e9d8c835653e39aa3c79428bf/packages/ariakit-react-core/src/utils/hooks.ts#L104-L113

Looks like the hook is available when using Preact, but it's noop. That's why it throws that error. The callback was never set.

**Comment by vio** [comment]
> I don't have plans to provide specific support for Preact. To do so, we'd need to have automated tests for Preact, which we've never had. If it's worked in the past, it's due to Preact's compatibility with React. So, if it's not working now, it's also due to Preact's compatibility with React.

Thank you for sharing, useful to know! Preact compatibility is fantastic for what it gives, though I had various compatibility issues with other libraries/frameworks. Aware that my case is not that common, I will try to keep these versions for a while and will reconsider.

> Looks like the hook is available when using Preact, but it's noop. That's why it throws that error. The callback was never set.

Thanks for pointing to it! I will try to have a deeper look, preact/compat uses useLayoutEffect,  will let you know if I find anything helpful.

**Comment by radimsv** [comment]
Hello everyone,
we are in front of a decision which, preferably small and maintained, component library to use in preact. I'm happy, somebody already tried :) 

Imho, it would really be great to have this library preact compatible, but I can imagine that without tests covering that case, it will become a nightmare sooner or later.

I am still not convinced that the `useInsertionEffect ` hook is used in the right way. As the docs state
> useInsertionEffect is for CSS-in-JS library authors. Unless you are working on a CSS-in-JS library and need a place to inject the styles, you probably want [useEffect](https://react.dev/reference/react/useEffect) or [useLayoutEffect](https://react.dev/reference/react/useLayoutEffect) instead

Is there a good reason why this hook is used @diegohaz ?

And you know, as Ariakit currently supports both react 17 and 18, it would probably make it easier for everybody to not use any hook from react@18 🤷 , i.e. use only the features available in the lowest supported version.

**Quick (not-really-a-)fix**
I tried to use `patch-package` to simulate that `preact` doesn't have the `useInsertionEffect` hook, and it looks better https://stackblitz.com/edit/gkmbyl-gxbbpf?file=vite.config.ts . `preact` doesn't have actual implementation for `useInsertionEffect` and it is just an alias for `useLayoutEffect` , [unfortunately](https://github.com/preactjs/preact/commit/b1cc8278f4adb0a501c54b71d7d1163804f6efab#diff-96d86484089f60cb07c0072f24b138dca9816de7176127d1b1b16fa8005a5d3cR33) - another patch could be to delete this hook from `preact`.
Unfortunately, the tooltip does not hide when the mouse is gone. I wasnt digging into this anymore... 

It's a pity. Although I really liked reakit, as the preact support is likely not to be added, or any mechanism to make the working components stay working in the future, it's probably not a good fit for preact.

**Comment by mayank99** [comment]
I was looking into this today and made some progress.

1. The first issue is the use of `useInsertionEffect` as mentioned above. Preact does shim `useInsertionEffect` as `useLayoutEffect`, but I think Ariakit (perhaps incorrectly) relies on the precise timing of React's lifecycle.
   - I worked around it by commenting out the `throw` statement. Commenting out the use of `useInsertionEffect` should also work fine.
2. After getting past the `useInsertionEffect` hurdle, the other issue was synthetic events which aren't present in Preact. It looks like [Ariakit mutates `event.currentTarget`](https://github.com/ariakit/ariakit/blob/d5471cf37a3efda2549685b19057a57ad4dfb68b/packages/ariakit-react-core/src/focusable/focusable.tsx#L330-L336) (again maybe a questionable thing to do). This isn't allowed in native DOM events, leading to this error: `Uncaught TypeError: Cannot set property currentTarget of #<Event> which has only a getter`.
   - I worked around it by using `Object.defineProperty` instead of directly assigning to `event.currentTarget`.

---

Here's a sandbox link: https://stackblitz.com/edit/vitejs-vite-9x2lcw?file=patches%2F%40ariakit%2Breact-core%2B0.4.15.patch,src%2Fapp.jsx&terminal=dev

<details>
<summary>@ariakit__react_core.patch</summary>

```diff
diff --git a/cjs/__chunks/KYD66ZXY.cjs b/cjs/__chunks/KYD66ZXY.cjs
index 3aa0b8192f1d0060c3fff20bb727c3bbe93869d3..5282a161a0bbed07680eab9e41487895f5ba99e8 100644
--- a/cjs/__chunks/KYD66ZXY.cjs
+++ b/cjs/__chunks/KYD66ZXY.cjs
@@ -235,7 +235,7 @@ var useFocusable = _6TI3NXX6cjs.createHook.call(void 0,
     });
     const handleFocusVisible = (event, currentTarget) => {
       if (currentTarget) {
-        event.currentTarget = currentTarget;
+        Object.defineProperty(event, 'currentTarget', {writable: false, value: currentTarget});
       }
       if (!focusable) return;
       const element = event.currentTarget;
diff --git a/cjs/__chunks/L5A2ID7S.cjs b/cjs/__chunks/L5A2ID7S.cjs
index 4df6065bba1652daf776ec567622a5398a2e428d..888e15872fcb7140bba3165429e65fbed0381eb5 100644
--- a/cjs/__chunks/L5A2ID7S.cjs
+++ b/cjs/__chunks/L5A2ID7S.cjs
@@ -51,9 +51,10 @@ function usePreviousValue(value) {
 }
 function useEvent(callback) {
   const ref = _react.useRef.call(void 0, () => {
-    throw new Error("Cannot call an event handler while rendering.");
+    console.error("Cannot call an event handler while rendering.");
   });
-  if (useReactInsertionEffect) {
+  // if (useReactInsertionEffect) {
+  if (false) {
     useReactInsertionEffect(() => {
       ref.current = callback;
     });
diff --git a/esm/__chunks/ABQUS43J.js b/esm/__chunks/ABQUS43J.js
index 04291224d1d2001747a820e9804a4fedb4a9d253..adac2f17104acd65bf3650ba7adf73a7b145f165 100644
--- a/esm/__chunks/ABQUS43J.js
+++ b/esm/__chunks/ABQUS43J.js
@@ -52,9 +52,10 @@ function usePreviousValue(value) {
 }
 function useEvent(callback) {
   const ref = useRef(() => {
-    throw new Error("Cannot call an event handler while rendering.");
+    console.error("Cannot call an event handler while rendering.");
   });
-  if (useReactInsertionEffect) {
+  // if (useReactInsertionEffect) {
+  if (false) {
     useReactInsertionEffect(() => {
       ref.current = callback;
     });
diff --git a/esm/__chunks/LVA2YJMS.js b/esm/__chunks/LVA2YJMS.js
index b3f0a6ffd311d099c4dc050976c5f55ca7d65422..60e31424ca93b99004863933e1de5c3e989d6135 100644
--- a/esm/__chunks/LVA2YJMS.js
+++ b/esm/__chunks/LVA2YJMS.js
@@ -236,7 +236,7 @@ var useFocusable = createHook(
     });
     const handleFocusVisible = (event, currentTarget) => {
       if (currentTarget) {
-        event.currentTarget = currentTarget;
+        Object.defineProperty(event, 'currentTarget', {writable: false, value: currentTarget})
       }
       if (!focusable) return;
       const element = event.currentTarget;

```

</details>

**Note**: I have only tested simple usage with a few different components and didn't see any issues.

---

If there is any interest in supporting Preact officially, I can help by contributing to the core packages and helping set up tests.

Aside from that, Ariakit could also help by providing unbundled source code so that we don't have to patch files named `__chunks/LVA2YJMS.js` which could change between versions.

**Reply by diegohaz** [reply]
> 1. The first issue is the use of `useInsertionEffect` as mentioned above. Preact does shim `useInsertionEffect` as `useLayoutEffect`, but I think Ariakit (perhaps incorrectly) relies on the precise timing of React's lifecycle.

It's a workaround because React doesn't yet provide a built-in `useEvent` equivalent function, but I don't think it's incorrect. For reference, [Bluesky uses the same approach](https://github.com/bluesky-social/social-app/blob/5f4a0f2881b9420f3a3f3fb6527352f58a99d9ea/src/lib/hooks/useNonReactiveCallback.ts#L13-L15) to achieve this goal.

I think Preact's decision to replace `useInsertionEffect` with `useLayoutEffect` is incorrect here.

> If there is any interest in supporting Preact officially, I can help by contributing to the core packages and helping set up tests.

While that would be nice, @benrodrs and I would need to dedicate a lot of time to it. Even if a community contributor agrees to help and maintain this part of the code, we would still have to handle Preact-specific issues with every change we make. I don't think it's fair to place all this burden on a community contributor.

This seems like a perfect use case for a fork.
